### PR TITLE
Implementing 'fit' for Torch

### PR DESCRIPTION
**Pull request by mjvypr**
We need to make a decisions whether `utils` will carry serialization/deserialization or `pubsub` will carry it.

TODO
- [ ] Create python Notebook to test Torch Integration

### CODE REVIEW

**Comment by mjvypr** [comment]
WIP

**Comment by jvmancuso** [comment]
Hey man, I was the one who put those TODOs and the serialization code in there.  Glad someone else is carrying on the Torch torch! (ha...)

I'd put the serialization code in utils.  That stuff was written before pubsub Grid was refactored into its current form.

Additionally, I'd planned on using [ignite](https://github.com/pytorch/ignite) for this version of fit.  This means the client will need to supply an explicit function for the training update for that stuff, and the worker will need to receive that and be able to run it.  That'll be tough.  There are a lot of reasons like this why Torch integration into Grid will be non-trivial.  I had several ideas, but I had higher priorities for Grid that took over, so if I can help at all feel free to reach out!

**Comment by mjvypr** [comment]
Interesting @jvmancuso I was thinking of this as my first serious commit ;) I will work with you for ignite, as I don't know a lot about that! (But will read). If it gets tough, I will try to clean up some code to start with!

**Comment by mjvypr** [comment]
Issue attempted: https://github.com/OpenMined/Grid/issues/59

**Comment by iamtrask** [comment]
Very excited about this PR! Love the progress so far @mjvypr !!!

**Comment by iamtrask** [comment]
(I intend to use this interface every day for my own personal research)

**Comment by iamtrask** [comment]
Hey @mjvypr how are things going? I'd love to help with this.

**Comment by bendecoste** [comment]
+1 serialization/deserialization in utils

**Comment by justin1121** [comment]
I'm pretty cautious about this feature. I think we need a proposed solution about how we're actually going to ship code to worker nodes because that could very dangerous. We either need a way to sandbox the training completely or somehow restrict which python code is called within the training loop. I think sandboxing would be the best way.

**Comment by bendecoste** [comment]
I have some security concerns in https://github.com/OpenMined/Grid/issues/48 as well that sandboxing could potentially address.

**Comment by bendecoste** [comment]
https://pypi.python.org/pypi/pysandbox/

**Comment by bendecoste** [comment]
One flow could be that the model creator creates their model and writes it to IPFS

A worker node downloads the model and saves it to ipfs
next, the worker kicks off a sandbox task, the task will
  - load the model
  - run the model
  - return gradients
outside the sandbox, we will create a new model using gradients
finish learning as it works right now

**Comment by bendecoste** [comment]
@justin1121 pointed out that the sandbox i linked is for python2 and kind of old .. I can't seem to find a good one for python 3

**Comment by jvmancuso** [comment]
We should be aware that even just loading model weights requires execution of arbitrary code because PyTorch uses Pickle, which has pretty strict requirements about directory structure and how packages are imported.  Just unpacking the pickle file has security risks.  Until PyTorch supports imports from ONNX, there's no trivial way to load a model without executing the code that defines the model class.  This is in contrast to Keras, where everything is packaged neatly in an hdf5.  Can we make it so that anytime you run a worker node, all of the code for Grid executes in a sandbox/VM?

**Comment by bendecoste** [comment]
>  Can we make it so that anytime you run a worker node, all of the code for Grid executes in a sandbox/VM?

Moved discussion about this to a separate issue because it is relevant to this and another issue.  See #66

**Comment by jvmancuso** [comment]
New approach to integrating PyTorch has begun in #104 that resolves security concerns and allows for more interactivity and flexibility.
